pyspider/fetcher/puppeteer_fetcher.js

```diff
@@ -0,0 +1,275 @@
+            script_result = await page.evaluate(fetch.js_script);
+            console.log("end script_result is: ", script_result);
+            script_executed = true;
+        }
+        page_content = await page.content();
+        end_time = Date.now() + wait_before_end;
+        setTimeout(make_result, wait_before_end + 10, page, http_response);
+    });
+
+    if (fetch.url && fetch.url.indexOf("://") == -1) {
+        fetch.url = "http://" + fetch.url;
+    }
+
+    width = fetch.js_viewport_width || 1024;
+    height = fetch.js_viewport_height || 768 * 3;
+    await page.setViewport({
+        "width": width,
+        "height": height
+    });
+
+    if (fetch.headers) {
+        fetch.headers = JSON.parse(fetch.headers)
+        await page.setExtraHTTPHeaders(fetch.headers);
+    }
+
+    if (fetch.headers && fetch.headers["User-Agent"]) {
+        console.log(fetch.headers["User-Agent"]);
+        console.log(typeof fetch.headers["User-Agent"]);
+        page.setUserAgent(fetch.headers["User-Agent"]);
+    }
+
+    var page_timeout = fetch.timeout ? fetch.timeout * 1000 : 20 * 1000;
+    await page.setDefaultNavigationTimeout(page_timeout);
+
+    async function make_result(page, http_response) {
+        if (finished) {
+            return;
+        }
+
+        if (Date.now() - start_time < page_timeout) {
+            if (!!!end_time) {
+                return;
+            }
+            if (end_time > Date.now()) {
+                setTimeout(make_result, Math.min(Date.now() - end_time, 100), page, http_response);
+                return;
+            }
+        }
+
+        finished = true;
+
+        var cookies = {};
+        var tmp_cookies = await page.cookies();
+        tmp_cookies.forEach(function (e) {
+            cookies[e.name] = e.value;
+        })
+
+        var page_content = await page.content();
+
+        var result = {};
+        try {
+            result = _make_result(page, page_content, cookies);
+            console.log("[" + result.status_code + "] " + result.orig_url + " " + result.time);
+        } catch (e) {
+            result = {
+                orig_url: fetch.url,
+                status_code: 599,
+                error: e.toString(),
+                content: page_content || "",
+                headers: {},
+                url: page.url() || fetch.url,
+                cookies: {},
+                time: (Date.now() - start_time) / 1000,
+                js_script_result: null,
+                save: fetch.save
+            }
+        }
+
+        var body = JSON.stringify(result, null, 4);
+
+        http_response.set({
+            "Cache": "no-cache",
+            "Content-Type": "application/json"
+        });
+
+        if (fetch.screenshot_path) {
+            await page.screenshot({path: fetch.screenshot_path});
+        }
+
+        http_response.send(body);
+
+        await page.close();
+        await browser.close();
+        return;
```

Comment:
spider as an async function, it will not exists after `make_result`, it will exit immediately after `page.goto`.
Your code may still work as it will run over the event loop, but it's not a good structure to combine async function and callback style code.

As async function is used here, you don't need to keep it like this. Just wait for page state change after `page.goto` then execute js, and so on then make result. (`Promise.race` may help)

Reply:
Thank you very much !I need it !